lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -333,6 +520,25 @@ actor ConnectorSink is Sink
       else
         Fail()
       end
+    | let sbt: CheckpointBarrierToken =>
+      if _message_processor.barrier_in_progress() then
+        _message_processor.receive_barrier(input_id, producer, barrier_token
+          where ack_barrier_if_complete = false)
+      else
+        try
+          // We don't want any messages to be processed by this sink
```

Comment:
Why are we queueing messages received over inputs we haven't received a barrier on yet?  I think there are only two consistent ways we handle this: (1) we process all messages prior to a checkpoint barrier (which includes all messages received over any input we haven't yet received the barrier on) or (2) we queue all messages at all times and only process the messages for a barrier once that checkpoint is complete.  This current approach seems to be halfway in between these.  

Note that it's non-deterministic which input we'll first receive a barrier over and when.  This means that which subset of messages for this checkpoint we queue rather than process by setting `always_queue = true` is also non-deterministic (and independent of where the barrier stands in each input stream).  The messages queued by the `BarrierSinkMessageProcessor` are treated as part of the next checkpoint, which means that here we're effectively partitioning the messages independently of the barrier, which violates assumptions underlying the barrier protocol.

Reply:
Another way to look at it is that when a Source is told it can ack upstream for a checkpoint id N, it assumes that means that all messages it sent before the barrier for N are finished processing.  But here you are potentially queueing messages it sent before the barrier for N and only processing them after checkpoint N, which undermines the at least once guarantee for that Source.

Reply:
This hunk of code may be leftover from a much earlier draft which did more, in error, in `process_barrier()`.  I'll work on a fix for this after lunch.